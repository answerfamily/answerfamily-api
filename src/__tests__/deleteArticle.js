jest.mock('../dataloaders/auth0UserLoaderFactory');

const {
  loadESFixtures,
  unloadESFixtures,
  loadMongoFixtures,
  unloadMongoFixtures,
} = require('../../test/fixtures');
const gql = require('../../test/gql');
const {
  ES_FIXTURES,
  MONGO_FIXTURES,
} = require('../__fixtures__/articleParagraphReply');
const mongoClient = require('../lib/mongoClient');
const esClient = require('../lib/esClient');

describe('deleteArticle', () => {
  beforeAll(async () => {
    await loadESFixtures(ES_FIXTURES);
    await loadMongoFixtures(MONGO_FIXTURES);
  });

  afterAll(async () => {
    await unloadESFixtures(ES_FIXTURES);
    await unloadMongoFixtures(MONGO_FIXTURES);
  });

  it('deletes article and associated paragraphs and paragraphReplies', async () => {
    const { errors, data } = await gql`
      mutation {
        deleteArticle(articleId: "article1")
      }
    `(
      {},
      {
        userPromise: Promise.resolve({ iss: 'author1' }),
      }
    );

    expect(errors).toBeUndefined();
    expect(data.deleteArticle).toBe(true);

    // Expect no article1 article
    expect(
      esClient.get({
        index: 'articles',
        type: '_doc',
        id: 'article1',
      })
    ).rejects.toMatchSnapshot('article not found error');

    // Expect no more article1 paragraphs
    const {
      hits: { total: paragraphTotalCount },
    } = await esClient.search({
      index: 'paragraphs',
      type: '_doc',
      body: {
        query: {
          term: {
            articleId: 'article1',
          },
        },
      },
    });
    expect(paragraphTotalCount).toBe(0);

    // Expect no more paragraph replies
    const { db } = await mongoClient;
    expect(
      await db
        .collection('paragraphReplies')
        .find({ paragraphId: { $in: ['p1', 'p2'] } })
        .toArray()
    ).toHaveLength(0);

    // cleanup -- add back deleted article, paragraph and paragraphReplies
    await esClient.index({
      index: 'articles',
      type: '_doc',
      id: 'article1',
      body: ES_FIXTURES['/articles/article1'],
    });

    await esClient.index({
      index: 'paragraphs',
      type: '_doc',
      id: 'p1',
      body: ES_FIXTURES['/paragraphs/p1'],
    });

    await esClient.index({
      index: 'paragraphs',
      type: '_doc',
      id: 'p2',
      body: ES_FIXTURES['/paragraphs/p2'],
    });

    await db
      .collection('paragraphReplies')
      .insertMany(MONGO_FIXTURES.paragraphReplies);
  });

  it('cannot delete article when author is not current user', async () => {
    const { errors } = await gql`
      mutation {
        deleteArticle(articleId: "article1")
      }
    `(
      {},
      {
        userPromise: Promise.resolve({ iss: 'another-user' }),
      }
    );

    // Should be forbidden
    expect(errors).toMatchSnapshot();
  });

  it('blocks non-logged in users', async () => {
    const { errors } = await gql`
      mutation {
        deleteArticle(articleId: "article1")
      }
    `();

    // Should be unauthorized
    expect(errors).toMatchSnapshot();
  });
});
