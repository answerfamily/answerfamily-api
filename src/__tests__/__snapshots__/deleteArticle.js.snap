// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deleteArticle blocks non-logged in users 1`] = `
Array [
  [GraphQLError: Must log in first],
]
`;

exports[`deleteArticle cannot delete article when author is not current user 1`] = `
Array [
  [GraphQLError: Only author can perform such action],
]
`;

exports[`deleteArticle deletes article and associated paragraphs and paragraphReplies: article not found error 1`] = `[Error: Not Found]`;
